---
title: "Raw sequencing data preprocessing"
date: '(`r format(Sys.time(), format = "%Y-%m-%d")`)'
output:
  flexdashboard::flex_dashboard:
    vertical_layout: scroll
    social: menu
    source_code: embed
    theme: cosmo
    self_contained: no
---

```{r setup, include = FALSE}
library(DT)
library(ggplot2)
library(purrr)
library(highcharter)
options(stringsAsFactors = F)
knitr::opts_chunk$set(echo = FALSE, 
                      message = FALSE,
                      warning = FALSE,
                      fig.align='center',fig.pos='H',
                      fig.path = "output_plots/",
                      dev = "svg")
```

<div id="preloader"></div>

```{js, echo=FALSE}
$(function() {
  $(window).load(function() {
    $('#preloader').fadeOut('slow',function(){$(this).remove();});
  });
});
```

```{css, echo=FALSE}
.limitrow {
  width: 99.4% !important;
  flex: none !important;
}

.limitrow .chart-stage {
  overflow-x:scroll;
}

#row-4 .chart-stage,
#row-5 .chart-stage {
width:1475px !important;
overflow-x:scroll !important;
}
```

Summary  {data-orientation=rows data-icon="fa-table"}
=====================================

### Fetch large-scale small RNA-seq data from NCBI and filter the raw data.
Users can collect public data from NCBI (https://www.ncbi.nlm.nih.gov/). Samples with low data quality will be excluded. The remaining samples were analyzed and indicated in the following tables and figures.

Row {data-width=850}
--------------------------------------

### Summary table {.limitrow}
```{r}
inputMat <- read.table("00Table_Summary_of_sRNA-seq_data.txt", stringsAsFactors = F, sep = "\t", row.names = 1)
inputMat <- inputMat[order(rownames(inputMat)), ]
adapter <- read.table("Adapter.txt", row.names = 1)
inputMat <- cbind(adapter[rownames(inputMat),], inputMat)
colnames(inputMat) <- c("Adapter", 'Phred', 'Barcode', "Raw reads", "Clean reads", "Ratio(%)", "Length limited", "Aligned", "Remove t/r/sn/snoRNA", "Collapsed reads","Most abundant read", "Most abundant read count")

for(tmp_nn in c(4,5,7,8,9,10,12)){
  inputMat[,tmp_nn] <- formatC(inputMat[,tmp_nn], big.mark=",", format="d")
}

datatable(data = inputMat, extensions = 'Buttons', class="compact cell-border", options = list( dom = "Blfrtip", buttons = list("copy",list(extend = "collection", buttons = c("csv", "excel", "pdf"), text = "Download") ),lengthMenu = list( c(10, 20, -1), c(10, 20, "All") ), pageLength = 10, autoWidth = TRUE), rownames = TRUE)%>%
formatStyle(TRUE, `text-align` = 'right')
```

<p style="font-size:14px;"> This table displays read number, adaptors, and barcode information for each library. `Adapter`: The adapter sequence; `Phred`: Phred score; `Barcode`: Whether there is a barcode in each library, "-" means no; `Raw reads`: Number of reads generated by sequencing; `Clean reads`: Number of reads that are filtered by adapters and quality scores; `Ratio`: Clean reads/Raw reads; `Length-limited`: Number of reads satisfying the length criteria; `Aligned`: Number of reads in limited length and at most one mismatch during mapping (-v 1 --best --strata); `Remove t/r/sn/snoRNA`: Number of mapped reads after removing reads corresponding to tRNAs, rRNAs, snRNAs and snoRNAs; `Collapsed reads`: Number of unique reads after reads collapsing; `Most abundant read`: The sequence with the highest count; `Most abundant read count`: The highest count. </p>

Row {data-width=850}
--------------------------------------

### Length distribution of all small RNAs

```{r}
Length_count <- read.table("Length_count.txt", sep = "\t")
Length_count <- Length_count[Length_count[,1]=="All", ]
Length_count <- Length_count[order(Length_count[,3]), ]
ds <- map(unique(Length_count[,2]), function(x){
    dt <- Length_count[Length_count[,2] == x, 3:4]
    dt <- list_parse2(as.data.frame(dt))
    list(data = dt, name = x)
})

highchart() %>% 
    hc_add_series_list(ds)%>%
    hc_add_theme(hc_theme_google())%>%
    hc_exporting(enabled = TRUE, filename = "Length distribution of collapsed small RNAs")

```

> The figure displays an interactive line chart which contains the length distribution of all reads in each small RNA-seq sample. The sample information is available by hovering over the points.

### Length distribution of collapsed small RNAs

```{r}
Length_count <- read.table("Length_count.txt", sep = "\t")
Length_count <- Length_count[Length_count[,1]=="Unique", ]
Length_count <- Length_count[order(Length_count[,3]), ]
ds <- map(unique(Length_count[,2]), function(x){
    dt <- Length_count[Length_count[,2] == x, 3:4]
    dt <- list_parse2(as.data.frame(dt))
    list(data = dt, name = x)
})

highchart() %>% 
    hc_add_series_list(ds)%>%
    hc_add_theme(hc_theme_google())%>%
    hc_exporting(enabled = TRUE, filename = "Length distribution of collapsed small RNAs")

```

> The figure displays an interactive line chart which contains the length distribution of unique reads in each small RNA-seq sample. The sample information is available by hovering over the points.
